web/pandas/pdeps/0009-io-extensions.md

```diff
@@ -0,0 +1,406 @@
+
+Third-party packages are already able to implement connectors to pandas, but
+there are some limitations to it:
+
+- Given the large number of formats supported by pandas itself, third-party
+  connectors are likely seen as second class citizens, not important enough
+  to be used, or not well supported.
+- There is no standard API for I/O connectors, and users of them need to learn
+  each of them individually.
+- Method chaining is not possible with third-party I/O connectors to export
+  data, unless authors monkey patch the `DataFrame` class, which should not
+  be encouraged.
+
+This document proposes to open the development of pandas I/O connectors to
+third-party libraries in a standard way that overcomes those limitations.
+
+### Proposal implementation
+
+Implementing this proposal would not require major changes to pandas, and
+the API defined next would be used.
+
+#### User API
+
+A new `.io` accessor would be created for the `DataFrame` class, where all
+I/O connector methods from third-parties would be loaded. Nothing else would
+live under that namespace. All methods in the `DataFrame.io` namespace would
+be consistently named `from_*` to import data from other sources to pandas
+and `to_*` to export data from pandas to other formats. For example:
+
+```python
+import pandas
+
+df = pandas.DataFrame.io.from_duckdb("SELECT * FROM 'dataset.parquet';")
+
+df.io.to_hive(hive_conn, "hive_table")
+```
+
+This API allows for method chaining:
+
+```python
+(pandas.DataFrame.io.from_duckdb("SELECT * FROM 'dataset.parquet';")
+                 .io.to_hive(hive_conn, "hive_table"))
+```
+
+By using a dedicated `.io` namespace, the amount of functions and methods in
+the already big namespaces of the `pandas` module and the `pandas.DataFrame`
+class will not be increased, and core pandas functionality would not be mixed
+with functionality registered via third-party plugins.
+
+#### Plugin registration
+
+Third-party packages would implement an
+[entrypoint](https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins)
+to define the connectors that they implement, under a group `dataframe.io`.
+
+For example, a hypothetical project `pandas_duckdb` implementing a `from_duckdb`
+function, could use `pyproject.toml` to define the next entry point:
+
+```toml
+[project.entry-points."dataframe.io"]
+from_duckdb = "pandas_duckdb:from_duckdb"
+```
+
+On import of the pandas module, it would read the entrypoint registry for the
+`dataframe.io` group, and would dynamically create methods in the `DataFrame.io`
+namespace for them. Method not starting with `from_` or `to_` would make pandas
+raise an exception. This would guarantee a reasonably consistent API among
+third-party connectors.
+
+#### Internal API
+
+Connectors would use Apache Arrow as the only interface to interact with pandas
+or any other project using them. In case an internal representation other than
```

Comment:
I'm struggling to follow this, sorry.

If converting between DataFrame libraries, then sure, this is clear

But what about the `from_csv` connector? Does this mean that would need to require arrow somewhere?

Perhaps if you spelled out what you envision what exactly you envision happening to `pandas.read_csv`, and where Arrow would come in, it would be clearer. Thanks!